contracts/extensions/StakedExpenditure.sol

```diff
@@ -0,0 +1,332 @@
+
+    emit StakeReclaimed(_expenditureId);
+  }
+
+  /// @notice Cancel the expenditure and reclaim the stake in one transaction
+  /// @notice Can only be called by expenditure owner while expenditure is in draft state
+  /// @param _permissionDomainId The domainId in which the extension has the arbitration permission
+  /// @param _childSkillIndex The index that the `_domainId` is relative to `_permissionDomainId`
+  /// @param _expenditureId The id of the expenditure
+  function cancelAndReclaimStake(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _expenditureId
+  )
+    public
+  {
+    Stake storage stake = stakes[_expenditureId];
+    ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(_expenditureId);
+
+    require(expenditure.owner == msgSender(), "staked-expenditure-must-be-owner");
+
+    require(
+      expenditure.status == ColonyDataTypes.ExpenditureStatus.Draft,
+      "staked-expenditure-expenditure-not-draft"
+    );
+
+    cancelExpenditure(_permissionDomainId, _childSkillIndex, _expenditureId, expenditure.owner);
+
+    // slither-disable-next-line reentrancy-no-eth
+    reclaimStake(_expenditureId);
+  }
+
+  /// @notice Cancel the expenditure and punish the staker
+  /// @notice Can only be called by an arbitration user
+  /// @param _permissionDomainId The domainId in which the extension has the arbitration permission
+  /// @param _childSkillIndex The index that the `_domainId` is relative to `_permissionDomainId`
+  /// @param _callerPermissionDomainId The domainId in which the caller has the arbitration permission
+  /// @param _callerChildSkillIndex The index that the `_domainId` is relative to `_callerPermissionDomainId`
+  /// @param _expenditureId The id of the expenditure
+  /// @param _punish Whether the staker should be punished by losing an amount of reputation equal to the stake
+  function cancelAndPunish(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _callerPermissionDomainId,
+    uint256 _callerChildSkillIndex,
+    uint256 _expenditureId,
+    bool _punish
+  )
+    public
+  {
+    ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(_expenditureId);
+
+    require(
+      colony.hasInheritedUserRole(
+        msgSender(),
+        _callerPermissionDomainId,
+        ColonyDataTypes.ColonyRole.Arbitration,
+        _callerChildSkillIndex,
+        expenditure.domainId
+      ),
+      "staked-expenditure-caller-not-arbitration"
+    );
+    require(
+      expenditure.status != ColonyDataTypes.ExpenditureStatus.Cancelled,
+      "staked-expenditure-expenditure-already-cancelled"
+    );
+
+    require(
+      expenditure.status != ColonyDataTypes.ExpenditureStatus.Draft,
+      "staked-expenditure-expenditure-still-draft"
+    );
+
+    if (_punish) {
+      Stake storage stake = stakes[_expenditureId];
+      require(stake.amount > 0, "staked-expenditure-nothing-to-slash");
```

Comment:
That's a good point. I would be inclined to delete the stake at the end of the function regardless, and only enter this branch if the amount is nonzero.

Reply:
After looking at this, deleting the stake regardless is no good since we need to support the flow where the user reclaims their stake after not being punished. I've made it that `_punish` deletes the stake regardless but only emits a penalty if the amount is nonzero.